// chromy/dist/index.js
'use strict';

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var fs = require('fs');

var CDP = require('chrome-remote-interface');
var chainProxy = require('async-chain-proxy');
var uuidV4 = require('uuid/v4');
var devices = require('./devices');
var sharp = require('sharp');

var _require = require('./emulation'),
    createFullscreenEmulationManager = _require.createFullscreenEmulationManager;

var _require2 = require('./error'),
    TimeoutError = _require2.TimeoutError,
    GotoTimeoutError = _require2.GotoTimeoutError,
    WaitTimeoutError = _require2.WaitTimeoutError,
    EvaluateTimeoutError = _require2.EvaluateTimeoutError,
    EvaluateError = _require2.EvaluateError;

var _require3 = require('./functionToSource'),
    functionToEvaluatingSource = _require3.functionToEvaluatingSource;

var _require4 = require('./util'),
    escapeHtml = _require4.escapeHtml,
    escapeSingleQuote = _require4.escapeSingleQuote,
    createChromeLauncher = _require4.createChromeLauncher,
    completeUrl = _require4.completeUrl;

var instances = [];
var instanceId = 1;

function makeSendToChromy(uuid) {
  return '\n  function () {\n    console.info(\'' + uuid + ':\' + JSON.stringify(arguments))\n  }\n  ';
}

function defaultTargetFunction(targets) {
  return targets.filter(function (t) {
    return t.type === 'page';
  }).shift();
}

var _clone = function _clone(obj) {
  return Object.assign({}, obj);
};

var Chromy = function () {
  function Chromy() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _classCallCheck3.default)(this, Chromy);

    var defaults = {
      port: 9222,
      waitTimeout: 30000,
      gotoTimeout: 30000,
      loadTimeout: 30000,
      evaluateTimeout: 30000,
      waitFunctionPollingInterval: 100,
      typeInterval: 20,
      activateOnStartUp: true,
      target: defaultTargetFunction,
      chromeFlags: []
    };
    this.options = Object.assign(_clone(defaults), options);
    this.cdpOptions = {
      port: this.options.port,
      target: this.options.target
    };
    this.client = null;
    this.launcher = null;
    this.messagePrefix = null;
    this.emulateMode = false;
    this.currentEmulateDeviceName = null;
    this.userAgentBeforeEmulate = null;
    this.instanceId = instanceId++;
  }

  (0, _createClass3.default)(Chromy, [{
    key: 'chain',
    value: function chain() {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      return chainProxy(this, options);
    }
  }, {
    key: 'start',
    value: function start() {
      var _this = this;

      var startingUrl = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
      return _regenerator2.default.async(function start$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              if (startingUrl === null) {
                startingUrl = 'about:blank';
              }

              if (!(this.client !== null)) {
                _context2.next = 3;
                break;
              }

              return _context2.abrupt('return');

            case 3:
              if (this.launcher === null) {
                this.launcher = createChromeLauncher(completeUrl(startingUrl), this.options);
              }
              _context2.next = 6;
              return _regenerator2.default.awrap(this.launcher.launch());

            case 6:
              instances.push(this);
              _context2.next = 9;
              return _regenerator2.default.awrap(new Promise(function (resolve, reject) {
                CDP(_this.cdpOptions, function _callee(client) {
                  var Network, Page, Runtime, Console, targetId;
                  return _regenerator2.default.async(function _callee$(_context) {
                    while (1) {
                      switch (_context.prev = _context.next) {
                        case 0:
                          _context.prev = 0;

                          _this.client = client;
                          Network = client.Network, Page = client.Page, Runtime = client.Runtime, Console = client.Console;
                          _context.next = 5;
                          return _regenerator2.default.awrap(Promise.all([Network.enable(), Page.enable(), Runtime.enable(), Console.enable()]));

                        case 5:
                          if (!_this.options.activateOnStartUp) {
                            _context.next = 11;
                            break;
                          }

                          _context.next = 8;
                          return _regenerator2.default.awrap(_this._getTargetIdFromOption());

                        case 8:
                          targetId = _context.sent;
                          _context.next = 11;
                          return _regenerator2.default.awrap(_this.client.Target.activateTarget({ targetId: targetId }));

                        case 11:
                          if (!('userAgent' in _this.options)) {
                            _context.next = 14;
                            break;
                          }

                          _context.next = 14;
                          return _regenerator2.default.awrap(_this.userAgent(_this.options.userAgent));

                        case 14:
                          if (!('headers' in _this.options)) {
                            _context.next = 17;
                            break;
                          }

                          _context.next = 17;
                          return _regenerator2.default.awrap(_this.headers(_this.options.headers));

                        case 17:
                          resolve(_this);
                          _context.next = 23;
                          break;

                        case 20:
                          _context.prev = 20;
                          _context.t0 = _context['catch'](0);

                          reject(_context.t0);

                        case 23:
                        case 'end':
                          return _context.stop();
                      }
                    }
                  }, null, _this, [[0, 20]]);
                }).on('error', function (err) {
                  reject(err);
                });
              }).catch(function (e) {
                throw e;
              }));

            case 9:
            case 'end':
              return _context2.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: '_getTargetIdFromOption',
    value: function _getTargetIdFromOption() {
      var result, page;
      return _regenerator2.default.async(function _getTargetIdFromOption$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              if (!(typeof this.options.target === 'function')) {
                _context3.next = 8;
                break;
              }

              _context3.next = 3;
              return _regenerator2.default.awrap(this.client.Target.getTargets());

            case 3:
              result = _context3.sent;
              page = this.options.target(result.targetInfos);
              return _context3.abrupt('return', page.targetId);

            case 8:
              if (!((0, _typeof3.default)(this.options.target) === 'object')) {
                _context3.next = 12;
                break;
              }

              return _context3.abrupt('return', this.options.target.targetId);

            case 12:
              if (!(typeof this.options.target === 'string')) {
                _context3.next = 16;
                break;
              }

              return _context3.abrupt('return', this.options.target);

            case 16:
              throw new Error('type of `target` option is invalid.');

            case 17:
            case 'end':
              return _context3.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'close',
    value: function close() {
      var _this2 = this;

      return _regenerator2.default.async(function close$(_context4) {
        while (1) {
          switch (_context4.prev = _context4.next) {
            case 0:
              if (!(this.client === null)) {
                _context4.next = 2;
                break;
              }

              return _context4.abrupt('return', false);

            case 2:
              _context4.next = 4;
              return _regenerator2.default.awrap(this.client.close());

            case 4:
              this.client = null;

              if (!(this.launcher !== null)) {
                _context4.next = 9;
                break;
              }

              _context4.next = 8;
              return _regenerator2.default.awrap(this.launcher.kill());

            case 8:
              this.launcher = null;

            case 9:
              instances = instances.filter(function (i) {
                return i.instanceId !== _this2.instanceId;
              });
              return _context4.abrupt('return', true);

            case 11:
            case 'end':
              return _context4.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'getPageTargets',
    value: function getPageTargets() {
      var result;
      return _regenerator2.default.async(function getPageTargets$(_context5) {
        while (1) {
          switch (_context5.prev = _context5.next) {
            case 0:
              _context5.next = 2;
              return _regenerator2.default.awrap(this.client.Target.getTargets());

            case 2:
              result = _context5.sent;
              return _context5.abrupt('return', result.targetInfos.filter(function (t) {
                return t.type === 'page';
              }));

            case 4:
            case 'end':
              return _context5.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'userAgent',
    value: function userAgent(ua) {
      return _regenerator2.default.async(function userAgent$(_context6) {
        while (1) {
          switch (_context6.prev = _context6.next) {
            case 0:
              _context6.next = 2;
              return _regenerator2.default.awrap(this._checkStart());

            case 2:
              _context6.next = 4;
              return _regenerator2.default.awrap(this.client.Network.setUserAgentOverride({ 'userAgent': ua }));

            case 4:
              return _context6.abrupt('return', _context6.sent);

            case 5:
            case 'end':
              return _context6.stop();
          }
        }
      }, null, this);
    }

    /**
     * Example:
     * chromy.headers({'X-Requested-By': 'foo'})
     */

  }, {
    key: 'headers',
    value: function headers(_headers) {
      return _regenerator2.default.async(function headers$(_context7) {
        while (1) {
          switch (_context7.prev = _context7.next) {
            case 0:
              _context7.next = 2;
              return _regenerator2.default.awrap(this._checkStart());

            case 2:
              _context7.next = 4;
              return _regenerator2.default.awrap(this.client.Network.setExtraHTTPHeaders({ 'headers': _headers }));

            case 4:
              return _context7.abrupt('return', _context7.sent);

            case 5:
            case 'end':
              return _context7.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'console',
    value: function (_console) {
      function console(_x) {
        return _console.apply(this, arguments);
      }

      console.toString = function () {
        return _console.toString();
      };

      return console;
    }(function _callee2(callback) {
      var _this3 = this;

      return _regenerator2.default.async(function _callee2$(_context8) {
        while (1) {
          switch (_context8.prev = _context8.next) {
            case 0:
              _context8.next = 2;
              return _regenerator2.default.awrap(this._checkStart());

            case 2:
              this.client.Console.messageAdded(function (payload) {
                try {
                  var msg = payload.message.text;
                  var pre = _this3.messagePrefix;
                  if (typeof msg !== 'undefined') {
                    if (pre === null || msg.substring(0, pre.length + 1) !== pre + ':') {
                      callback.apply(_this3, [msg, payload.message]);
                    }
                  }
                } catch (e) {
                  console.warn(e);
                }
              });

            case 3:
            case 'end':
              return _context8.stop();
          }
        }
      }, null, this);
    })
  }, {
    key: 'receiveMessage',
    value: function receiveMessage(callback) {
      var _this4 = this;

      var uuid, f;
      return _regenerator2.default.async(function receiveMessage$(_context9) {
        while (1) {
          switch (_context9.prev = _context9.next) {
            case 0:
              _context9.next = 2;
              return _regenerator2.default.awrap(this._checkStart());

            case 2:
              uuid = uuidV4();

              this.messagePrefix = uuid;
              f = makeSendToChromy(this.messagePrefix);

              this.defineFunction({ sendToChromy: f });
              this.client.Console.messageAdded(function (payload) {
                try {
                  var msg = payload.message.text;
                  if (msg && msg.substring(0, uuid.length + 1) === uuid + ':') {
                    var data = JSON.parse(msg.substring(uuid.length + 1));
                    callback.apply(_this4, [data]);
                  }
                } catch (e) {
                  console.warn(e);
                }
              });

            case 7:
            case 'end':
              return _context9.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'goto',
    value: function goto(url, options) {
      var _this5 = this;

      var defaultOptions;
      return _regenerator2.default.async(function goto$(_context11) {
        while (1) {
          switch (_context11.prev = _context11.next) {
            case 0:
              defaultOptions = {
                waitLoadEvent: true
              };

              options = Object.assign(_clone(defaultOptions), options);
              _context11.next = 4;
              return _regenerator2.default.awrap(this._checkStart(url));

            case 4:
              _context11.prev = 4;
              _context11.next = 7;
              return _regenerator2.default.awrap(this._waitFinish(this.options.gotoTimeout, function _callee3() {
                return _regenerator2.default.async(function _callee3$(_context10) {
                  while (1) {
                    switch (_context10.prev = _context10.next) {
                      case 0:
                        _context10.next = 2;
                        return _regenerator2.default.awrap(_this5.client.Page.navigate({ url: completeUrl(url) }));

                      case 2:
                        if (!options.waitLoadEvent) {
                          _context10.next = 5;
                          break;
                        }

                        _context10.next = 5;
                        return _regenerator2.default.awrap(_this5.client.Page.loadEventFired());

                      case 5:
                      case 'end':
                        return _context10.stop();
                    }
                  }
                }, null, _this5);
              }));

            case 7:
              _context11.next = 16;
              break;

            case 9:
              _context11.prev = 9;
              _context11.t0 = _context11['catch'](4);

              if (!(_context11.t0 instanceof TimeoutError)) {
                _context11.next = 15;
                break;
              }

              throw new GotoTimeoutError('goto() timeout');

            case 15:
              throw _context11.t0;

            case 16:
            case 'end':
              return _context11.stop();
          }
        }
      }, null, this, [[4, 9]]);
    }
  }, {
    key: 'waitLoadEvent',
    value: function waitLoadEvent() {
      var _this6 = this;

      return _regenerator2.default.async(function waitLoadEvent$(_context13) {
        while (1) {
          switch (_context13.prev = _context13.next) {
            case 0:
              _context13.next = 2;
              return _regenerator2.default.awrap(this._waitFinish(this.options.loadTimeout, function _callee4() {
                return _regenerator2.default.async(function _callee4$(_context12) {
                  while (1) {
                    switch (_context12.prev = _context12.next) {
                      case 0:
                        _context12.next = 2;
                        return _regenerator2.default.awrap(_this6.client.Page.loadEventFired());

                      case 2:
                      case 'end':
                        return _context12.stop();
                    }
                  }
                }, null, _this6);
              }));

            case 2:
            case 'end':
              return _context13.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'forward',
    value: function forward() {
      var f, promise;
      return _regenerator2.default.async(function forward$(_context14) {
        while (1) {
          switch (_context14.prev = _context14.next) {
            case 0:
              f = 'window.history.forward()';
              promise = this.waitLoadEvent();
              _context14.next = 4;
              return _regenerator2.default.awrap(this.client.Runtime.evaluate({ expression: f }));

            case 4:
              _context14.next = 6;
              return _regenerator2.default.awrap(promise);

            case 6:
            case 'end':
              return _context14.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'back',
    value: function back() {
      var f, promise;
      return _regenerator2.default.async(function back$(_context15) {
        while (1) {
          switch (_context15.prev = _context15.next) {
            case 0:
              f = 'window.history.back()';
              promise = this.waitLoadEvent();
              _context15.next = 4;
              return _regenerator2.default.awrap(this.client.Runtime.evaluate({ expression: f }));

            case 4:
              _context15.next = 6;
              return _regenerator2.default.awrap(promise);

            case 6:
            case 'end':
              return _context15.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'reload',
    value: function reload(ignoreCache, scriptToEvaluateOnLoad) {
      return _regenerator2.default.async(function reload$(_context16) {
        while (1) {
          switch (_context16.prev = _context16.next) {
            case 0:
              _context16.next = 2;
              return _regenerator2.default.awrap(this.client.Page.reload({ ignoreCache: ignoreCache, scriptToEvaluateOnLoad: scriptToEvaluateOnLoad }));

            case 2:
            case 'end':
              return _context16.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'evaluate',
    value: function evaluate(expr) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      return _regenerator2.default.async(function evaluate$(_context17) {
        while (1) {
          switch (_context17.prev = _context17.next) {
            case 0:
              _context17.next = 2;
              return _regenerator2.default.awrap(this._evaluateWithReplaces(expr, options));

            case 2:
              return _context17.abrupt('return', _context17.sent);

            case 3:
            case 'end':
              return _context17.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: '_evaluateWithReplaces',
    value: function _evaluateWithReplaces(expr) {
      var _this7 = this;

      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var replaces = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      var e, result, resultObject, type;
      return _regenerator2.default.async(function _evaluateWithReplaces$(_context19) {
        while (1) {
          switch (_context19.prev = _context19.next) {
            case 0:
              e = functionToEvaluatingSource(expr, replaces);
              _context19.prev = 1;
              _context19.next = 4;
              return _regenerator2.default.awrap(this._waitFinish(this.options.evaluateTimeout, function _callee5() {
                var params;
                return _regenerator2.default.async(function _callee5$(_context18) {
                  while (1) {
                    switch (_context18.prev = _context18.next) {
                      case 0:
                        if (_this7.client) {
                          _context18.next = 2;
                          break;
                        }

                        return _context18.abrupt('return', null);

                      case 2:
                        params = Object.assign({}, options, { expression: e });
                        _context18.next = 5;
                        return _regenerator2.default.awrap(_this7.client.Runtime.evaluate(params));

                      case 5:
                        return _context18.abrupt('return', _context18.sent);

                      case 6:
                      case 'end':
                        return _context18.stop();
                    }
                  }
                }, null, _this7);
              }));

            case 4:
              result = _context19.sent;

              if (!(!result || !result.result)) {
                _context19.next = 7;
                break;
              }

              return _context19.abrupt('return', null);

            case 7:
              if (!(result.result.subtype === 'promise')) {
                _context19.next = 12;
                break;
              }

              _context19.next = 10;
              return _regenerator2.default.awrap(this.client.Runtime.awaitPromise({ promiseObjectId: result.result.objectId, returnByValue: true }));

            case 10:
              result = _context19.sent;

              // adjust to after process
              result.result.value = JSON.stringify({
                type: (0, _typeof3.default)(result.result.value),
                result: JSON.stringify(result.result.value)
              });

            case 12:
              if (!(result.result.subtype === 'error')) {
                _context19.next = 14;
                break;
              }

              throw new EvaluateError('An error has been occurred in evaluated script on a browser.' + result.result.description, result.result);

            case 14:
              resultObject = JSON.parse(result.result.value);
              type = resultObject.type;

              if (!(type === 'undefined')) {
                _context19.next = 20;
                break;
              }

              return _context19.abrupt('return', undefined);

            case 20:
              _context19.prev = 20;
              return _context19.abrupt('return', JSON.parse(resultObject.result));

            case 24:
              _context19.prev = 24;
              _context19.t0 = _context19['catch'](20);

              console.log('ERROR', resultObject);
              throw _context19.t0;

            case 28:
              _context19.next = 37;
              break;

            case 30:
              _context19.prev = 30;
              _context19.t1 = _context19['catch'](1);

              if (!(_context19.t1 instanceof TimeoutError)) {
                _context19.next = 36;
                break;
              }

              throw new EvaluateTimeoutError('evaluate() timeout');

            case 36:
              throw _context19.t1;

            case 37:
            case 'end':
              return _context19.stop();
          }
        }
      }, null, this, [[1, 30], [20, 24]]);
    }
  }, {
    key: '_waitFinish',
    value: function _waitFinish(timeout, callback) {
      var _this8 = this;

      var start, finished, error, result, f, now;
      return _regenerator2.default.async(function _waitFinish$(_context21) {
        while (1) {
          switch (_context21.prev = _context21.next) {
            case 0:
              start = Date.now();
              finished = false;
              error = null;
              result = null;

              f = function _callee6() {
                return _regenerator2.default.async(function _callee6$(_context20) {
                  while (1) {
                    switch (_context20.prev = _context20.next) {
                      case 0:
                        _context20.prev = 0;
                        _context20.next = 3;
                        return _regenerator2.default.awrap(callback.apply());

                      case 3:
                        result = _context20.sent;

                        finished = true;
                        return _context20.abrupt('return', result);

                      case 8:
                        _context20.prev = 8;
                        _context20.t0 = _context20['catch'](0);

                        error = _context20.t0;
                        finished = true;

                      case 12:
                      case 'end':
                        return _context20.stop();
                    }
                  }
                }, null, _this8, [[0, 8]]);
              };

              f.apply();

            case 6:
              if (finished) {
                _context21.next = 14;
                break;
              }

              now = Date.now();

              if (!(now - start > timeout)) {
                _context21.next = 10;
                break;
              }

              throw new TimeoutError('timeout');

            case 10:
              _context21.next = 12;
              return _regenerator2.default.awrap(this.sleep(this.options.waitFunctionPollingInterval));

            case 12:
              _context21.next = 6;
              break;

            case 14:
              if (!(error !== null)) {
                _context21.next = 16;
                break;
              }

              throw error;

            case 16:
              return _context21.abrupt('return', result);

            case 17:
            case 'end':
              return _context21.stop();
          }
        }
      }, null, this);
    }

    /**
     * define function
     *
     * @param func {(function|string|Array.<function>|Array.<string>)}
     * @returns {Promise.<void>}
     */

  }, {
    key: 'defineFunction',
    value: function defineFunction(def) {
      var funcs, i, f;
      return _regenerator2.default.async(function defineFunction$(_context22) {
        while (1) {
          switch (_context22.prev = _context22.next) {
            case 0:
              funcs = [];

              if (Array.isArray(def)) {
                funcs = def;
              } else if ((typeof def === 'undefined' ? 'undefined' : (0, _typeof3.default)(def)) === 'object') {
                funcs = this._moduleToFunctionSources(def);
              } else {
                funcs.push(def);
              }
              i = 0;

            case 3:
              if (!(i < funcs.length)) {
                _context22.next = 11;
                break;
              }

              f = funcs[i];

              if (typeof f === 'function') {
                f = f.toString();
              }
              _context22.next = 8;
              return _regenerator2.default.awrap(this.client.Runtime.evaluate({ expression: f }));

            case 8:
              i++;
              _context22.next = 3;
              break;

            case 11:
            case 'end':
              return _context22.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: '_moduleToFunctionSources',
    value: function _moduleToFunctionSources(module) {
      var result = [];
      for (var funcName in module) {
        var func = module[funcName];
        var src = ('function ' + funcName + ' () { return (' + func.toString() + ')(...arguments) }').trim();
        result.push(src);
      }
      return result;
    }
  }, {
    key: 'sleep',
    value: function sleep(msec) {
      return _regenerator2.default.async(function sleep$(_context23) {
        while (1) {
          switch (_context23.prev = _context23.next) {
            case 0:
              _context23.next = 2;
              return _regenerator2.default.awrap(new Promise(function (resolve, reject) {
                setTimeout(function () {
                  resolve();
                }, msec);
              }));

            case 2:
            case 'end':
              return _context23.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'wait',
    value: function wait(cond) {
      return _regenerator2.default.async(function wait$(_context24) {
        while (1) {
          switch (_context24.prev = _context24.next) {
            case 0:
              if (!(typeof cond === 'number')) {
                _context24.next = 5;
                break;
              }

              _context24.next = 3;
              return _regenerator2.default.awrap(this.sleep(cond));

            case 3:
              _context24.next = 12;
              break;

            case 5:
              if (!(typeof cond === 'function')) {
                _context24.next = 10;
                break;
              }

              _context24.next = 8;
              return _regenerator2.default.awrap(this._waitFunction(cond));

            case 8:
              _context24.next = 12;
              break;

            case 10:
              _context24.next = 12;
              return _regenerator2.default.awrap(this._waitSelector(cond));

            case 12:
            case 'end':
              return _context24.stop();
          }
        }
      }, null, this);
    }

    // wait for func to return true.

  }, {
    key: '_waitFunction',
    value: function _waitFunction(func) {
      var _this9 = this;

      return _regenerator2.default.async(function _waitFunction$(_context26) {
        while (1) {
          switch (_context26.prev = _context26.next) {
            case 0:
              _context26.next = 2;
              return _regenerator2.default.awrap(this._waitFinish(this.options.waitTimeout, function _callee7() {
                var r;
                return _regenerator2.default.async(function _callee7$(_context25) {
                  while (1) {
                    switch (_context25.prev = _context25.next) {
                      case 0:
                        if (!true) {
                          _context25.next = 10;
                          break;
                        }

                        _context25.next = 3;
                        return _regenerator2.default.awrap(_this9.evaluate(func));

                      case 3:
                        r = _context25.sent;

                        if (!r) {
                          _context25.next = 6;
                          break;
                        }

                        return _context25.abrupt('break', 10);

                      case 6:
                        _context25.next = 8;
                        return _regenerator2.default.awrap(_this9.sleep(_this9.options.waitFunctionPollingInterval));

                      case 8:
                        _context25.next = 0;
                        break;

                      case 10:
                      case 'end':
                        return _context25.stop();
                    }
                  }
                }, null, _this9);
              }));

            case 2:
            case 'end':
              return _context26.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: '_waitSelector',
    value: function _waitSelector(selector) {
      var _this10 = this;

      var _check, startTime;

      return _regenerator2.default.async(function _waitSelector$(_context28) {
        while (1) {
          switch (_context28.prev = _context28.next) {
            case 0:
              _check = null;
              startTime = Date.now();
              _context28.next = 4;
              return _regenerator2.default.awrap(new Promise(function (resolve, reject) {
                _check = function check() {
                  setTimeout(function _callee8() {
                    var now, result;
                    return _regenerator2.default.async(function _callee8$(_context27) {
                      while (1) {
                        switch (_context27.prev = _context27.next) {
                          case 0:
                            _context27.prev = 0;
                            now = Date.now();

                            if (!(now - startTime > _this10.options.waitTimeout)) {
                              _context27.next = 5;
                              break;
                            }

                            reject(new WaitTimeoutError('wait() timeout'));
                            return _context27.abrupt('return');

                          case 5:
                            _context27.next = 7;
                            return _regenerator2.default.awrap(_this10._evaluateWithReplaces(function () {
                              return document.querySelector('?');
                            }, {}, { '?': escapeSingleQuote(selector) }));

                          case 7:
                            result = _context27.sent;

                            if (result) {
                              resolve(result);
                            } else {
                              _check();
                            }
                            _context27.next = 14;
                            break;

                          case 11:
                            _context27.prev = 11;
                            _context27.t0 = _context27['catch'](0);

                            reject(_context27.t0);

                          case 14:
                          case 'end':
                            return _context27.stop();
                        }
                      }
                    }, null, _this10, [[0, 11]]);
                  }, _this10.options.waitFunctionPollingInterval);
                };
                _check();
              }));

            case 4:
            case 'end':
              return _context28.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'type',
    value: function type(expr, value) {
      var characters, i, c;
      return _regenerator2.default.async(function type$(_context29) {
        while (1) {
          switch (_context29.prev = _context29.next) {
            case 0:
              _context29.next = 2;
              return _regenerator2.default.awrap(this.evaluate('document.querySelector("' + expr + '").focus()'));

            case 2:
              characters = value.split('');
              _context29.t0 = _regenerator2.default.keys(characters);

            case 4:
              if ((_context29.t1 = _context29.t0()).done) {
                _context29.next = 13;
                break;
              }

              i = _context29.t1.value;
              c = characters[i];
              _context29.next = 9;
              return _regenerator2.default.awrap(this.client.Input.dispatchKeyEvent({ type: 'char', text: c }));

            case 9:
              _context29.next = 11;
              return _regenerator2.default.awrap(this.sleep(this.options.typeInterval));

            case 11:
              _context29.next = 4;
              break;

            case 13:
            case 'end':
              return _context29.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'insert',
    value: function insert(expr, value) {
      return _regenerator2.default.async(function insert$(_context30) {
        while (1) {
          switch (_context30.prev = _context30.next) {
            case 0:
              expr = escapeSingleQuote(expr);
              _context30.next = 3;
              return _regenerator2.default.awrap(this.evaluate('document.querySelector(\'' + expr + '\').focus()'));

            case 3:
              _context30.next = 5;
              return _regenerator2.default.awrap(this.evaluate('document.querySelector(\'' + expr + '\').value = "' + escapeHtml(value) + '"'));

            case 5:
            case 'end':
              return _context30.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'click',
    value: function click(expr) {
      var inputOptions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var defaults, options, promise;
      return _regenerator2.default.async(function click$(_context31) {
        while (1) {
          switch (_context31.prev = _context31.next) {
            case 0:
              defaults = { waitLoadEvent: false };
              options = Object.assign(_clone(defaults), inputOptions);
              promise = null;

              if (options.waitLoadEvent) {
                promise = this.waitLoadEvent();
              }
              _context31.next = 6;
              return _regenerator2.default.awrap(this.evaluate('document.querySelectorAll(\'' + escapeSingleQuote(expr) + '\').forEach(n => n.click())'));

            case 6:
              if (!(promise !== null)) {
                _context31.next = 9;
                break;
              }

              _context31.next = 9;
              return _regenerator2.default.awrap(promise);

            case 9:
            case 'end':
              return _context31.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'mouseMoved',
    value: function mouseMoved(x, y) {
      var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      var opts;
      return _regenerator2.default.async(function mouseMoved$(_context32) {
        while (1) {
          switch (_context32.prev = _context32.next) {
            case 0:
              opts = Object.assign({ type: 'mouseMoved', x: x, y: y }, options);
              _context32.next = 3;
              return _regenerator2.default.awrap(this.client.Input.dispatchMouseEvent(opts));

            case 3:
            case 'end':
              return _context32.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'mousePressed',
    value: function mousePressed(x, y) {
      var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      var opts;
      return _regenerator2.default.async(function mousePressed$(_context33) {
        while (1) {
          switch (_context33.prev = _context33.next) {
            case 0:
              opts = Object.assign({ type: 'mousePressed', x: x, y: y, button: 'left' }, options);
              _context33.next = 3;
              return _regenerator2.default.awrap(this.client.Input.dispatchMouseEvent(opts));

            case 3:
            case 'end':
              return _context33.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'mouseReleased',
    value: function mouseReleased(x, y) {
      var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      var opts;
      return _regenerator2.default.async(function mouseReleased$(_context34) {
        while (1) {
          switch (_context34.prev = _context34.next) {
            case 0:
              opts = Object.assign({ type: 'mouseReleased', x: x, y: y, button: 'left' }, options);
              _context34.next = 3;
              return _regenerator2.default.awrap(this.client.Input.dispatchMouseEvent(opts));

            case 3:
            case 'end':
              return _context34.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'tap',
    value: function tap(x, y) {
      var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      var time, opts;
      return _regenerator2.default.async(function tap$(_context35) {
        while (1) {
          switch (_context35.prev = _context35.next) {
            case 0:
              time = Date.now() / 1000;
              opts = Object.assign({ x: x, y: y, timestamp: time, button: 'left' }, options);
              _context35.next = 4;
              return _regenerator2.default.awrap(this.client.Input.synthesizeTapGesture(opts));

            case 4:
            case 'end':
              return _context35.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'doubleTap',
    value: function doubleTap(x, y) {
      var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      var time, opts;
      return _regenerator2.default.async(function doubleTap$(_context36) {
        while (1) {
          switch (_context36.prev = _context36.next) {
            case 0:
              time = Date.now() / 1000;
              opts = Object.assign({ x: x, y: y, timestamp: time, button: 'left', tapCount: 2 }, options);
              _context36.next = 4;
              return _regenerator2.default.awrap(this.client.Input.synthesizeTapGesture(opts));

            case 4:
            case 'end':
              return _context36.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'check',
    value: function check(selector) {
      return _regenerator2.default.async(function check$(_context37) {
        while (1) {
          switch (_context37.prev = _context37.next) {
            case 0:
              _context37.next = 2;
              return _regenerator2.default.awrap(this.evaluate('document.querySelectorAll(\'' + escapeSingleQuote(selector) + '\').forEach(n => n.checked = true)'));

            case 2:
            case 'end':
              return _context37.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'uncheck',
    value: function uncheck(selector) {
      return _regenerator2.default.async(function uncheck$(_context38) {
        while (1) {
          switch (_context38.prev = _context38.next) {
            case 0:
              _context38.next = 2;
              return _regenerator2.default.awrap(this.evaluate('document.querySelectorAll(\'' + escapeSingleQuote(selector) + '\').forEach(n => n.checked = false)'));

            case 2:
            case 'end':
              return _context38.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'select',
    value: function select(selector, value) {
      var sel, src;
      return _regenerator2.default.async(function select$(_context39) {
        while (1) {
          switch (_context39.prev = _context39.next) {
            case 0:
              sel = escapeSingleQuote(selector);
              src = '\n      document.querySelectorAll(\'' + sel + ' > option\').forEach(n => {\n        if (n.value === "' + value + '") {\n          n.selected = true\n        }\n      })\n      ';
              _context39.next = 4;
              return _regenerator2.default.awrap(this.evaluate(src));

            case 4:
            case 'end':
              return _context39.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'screenshot',
    value: function screenshot() {
      var format = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'png';
      var quality = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : undefined;
      var fromSurface = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

      var _ref, data;

      return _regenerator2.default.async(function screenshot$(_context40) {
        while (1) {
          switch (_context40.prev = _context40.next) {
            case 0:
              if (!(['png', 'jpeg'].indexOf(format) === -1)) {
                _context40.next = 2;
                break;
              }

              throw new Error('format is invalid.');

            case 2:
              _context40.next = 4;
              return _regenerator2.default.awrap(this.client.Page.captureScreenshot({
                format: format,
                quality: quality,
                fromSurface: fromSurface
              }));

            case 4:
              _ref = _context40.sent;
              data = _ref.data;
              return _context40.abrupt('return', Buffer.from(data, 'base64'));

            case 7:
            case 'end':
              return _context40.stop();
          }
        }
      }, null, this);
    }

    /*
     * Limitation:
     * maximum height is 16384px because of chrome's bug from Skia library.
     * https://groups.google.com/a/chromium.org/d/msg/headless-dev/DqaAEXyzvR0/kUTEqNYiDQAJ
     * https://stackoverflow.com/questions/44599858/max-height-of-16-384px-for-headless-chrome-screenshots
     */

  }, {
    key: 'screenshotDocument',
    value: function screenshotDocument() {
      var model = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'scroll';
      var format = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'png';
      var quality = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : undefined;
      var fromSurface = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : true;
      var emulation, result, info, s, m1, newWidth, newHeight;
      return _regenerator2.default.async(function screenshotDocument$(_context41) {
        while (1) {
          switch (_context41.prev = _context41.next) {
            case 0:
              _context41.next = 2;
              return _regenerator2.default.awrap(createFullscreenEmulationManager(this, model));

            case 2:
              emulation = _context41.sent;
              result = null;
              _context41.prev = 4;
              _context41.next = 7;
              return _regenerator2.default.awrap(emulation.emulate());

            case 7:
              _context41.next = 9;
              return _regenerator2.default.awrap(this.screenshot(format, quality, fromSurface));

            case 9:
              result = _context41.sent;
              info = emulation.browserInfo;

              if (!(info.devicePixelRatio !== 1)) {
                _context41.next = 21;
                break;
              }

              s = sharp(result);
              _context41.next = 15;
              return _regenerator2.default.awrap(s.metadata());

            case 15:
              m1 = _context41.sent;
              newWidth = parseInt(m1.width / info.devicePixelRatio);
              newHeight = parseInt(m1.height / info.devicePixelRatio);
              _context41.next = 20;
              return _regenerator2.default.awrap(s.resize(newWidth, newHeight).toBuffer());

            case 20:
              result = _context41.sent;

            case 21:
              _context41.prev = 21;
              _context41.next = 24;
              return _regenerator2.default.awrap(emulation.reset());

            case 24:
              if (!(this.currentEmulateDeviceName !== null)) {
                _context41.next = 27;
                break;
              }

              _context41.next = 27;
              return _regenerator2.default.awrap(this.emulate(this.currentEmulateDeviceName));

            case 27:
              return _context41.finish(21);

            case 28:
              return _context41.abrupt('return', result);

            case 29:
            case 'end':
              return _context41.stop();
          }
        }
      }, null, this, [[4,, 21, 28]]);
    }
  }, {
    key: 'screenshotSelector',
    value: function screenshotSelector(selector) {
      var format = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'png';
      var quality = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : undefined;
      var fromSurface = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : true;
      var rect, pixelRatio, actualRect, clipRect, buffer, meta;
      return _regenerator2.default.async(function screenshotSelector$(_context42) {
        while (1) {
          switch (_context42.prev = _context42.next) {
            case 0:
              _context42.next = 2;
              return _regenerator2.default.awrap(this.getBoundingClientRect(selector));

            case 2:
              rect = _context42.sent;

              if (rect) {
                _context42.next = 5;
                break;
              }

              return _context42.abrupt('return', null);

            case 5:
              _context42.next = 7;
              return _regenerator2.default.awrap(this.evaluate(function () {
                return window.devicePixelRatio;
              }));

            case 7:
              pixelRatio = _context42.sent;
              _context42.next = 10;
              return _regenerator2.default.awrap(this.scroll(rect.left, rect.top));

            case 10:
              _context42.next = 12;
              return _regenerator2.default.awrap(this.getBoundingClientRect(selector));

            case 12:
              actualRect = _context42.sent;

              if (actualRect) {
                _context42.next = 15;
                break;
              }

              return _context42.abrupt('return', null);

            case 15:
              clipRect = {
                top: Math.floor(actualRect.top * pixelRatio),
                left: Math.floor(actualRect.left * pixelRatio),
                width: Math.floor(actualRect.width * pixelRatio),
                height: Math.floor(actualRect.height * pixelRatio)
              };
              _context42.next = 18;
              return _regenerator2.default.awrap(this.screenshot(format, quality, fromSurface));

            case 18:
              buffer = _context42.sent;
              _context42.next = 21;
              return _regenerator2.default.awrap(sharp(buffer).metadata());

            case 21:
              meta = _context42.sent;

              if (meta.width < clipRect.left + clipRect.width) {
                clipRect.width = meta.width - clipRect.left;
              }
              if (meta.height < clipRect.top + clipRect.height) {
                clipRect.height = meta.height - clipRect.top;
              }
              return _context42.abrupt('return', sharp(buffer).extract(clipRect).toBuffer());

            case 25:
            case 'end':
              return _context42.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'screenshotMultipleSelectors',
    value: function screenshotMultipleSelectors(selectors, callback) {
      var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

      var defaults, opts, fullscreenBuffer, meta, emulation, selIdx, selector, rects, r, err, rectIdx, rect, _err, buffer;

      return _regenerator2.default.async(function screenshotMultipleSelectors$(_context43) {
        while (1) {
          switch (_context43.prev = _context43.next) {
            case 0:
              defaults = {
                model: 'scroll',
                format: 'png',
                quality: undefined,
                fromSurface: true,
                useSelectorAll: false
              };
              opts = Object.assign({}, defaults, options);
              _context43.next = 4;
              return _regenerator2.default.awrap(this.screenshotDocument(opts.model, opts.format, opts.quality, opts.fromSurface));

            case 4:
              fullscreenBuffer = _context43.sent;
              _context43.next = 7;
              return _regenerator2.default.awrap(sharp(fullscreenBuffer).metadata());

            case 7:
              meta = _context43.sent;
              _context43.next = 10;
              return _regenerator2.default.awrap(createFullscreenEmulationManager(this, 'scroll'));

            case 10:
              emulation = _context43.sent;
              _context43.next = 13;
              return _regenerator2.default.awrap(emulation.emulate());

            case 13:
              _context43.prev = 13;
              selIdx = 0;

            case 15:
              if (!(selIdx < selectors.length)) {
                _context43.next = 57;
                break;
              }

              selector = selectors[selIdx];
              _context43.prev = 17;
              rects = null;

              if (!opts.useSelectorAll) {
                _context43.next = 25;
                break;
              }

              _context43.next = 22;
              return _regenerator2.default.awrap(this.getBoundingClientRectAll(selector));

            case 22:
              rects = _context43.sent;
              _context43.next = 29;
              break;

            case 25:
              _context43.next = 27;
              return _regenerator2.default.awrap(this.getBoundingClientRect(selector));

            case 27:
              r = _context43.sent;

              if (r) {
                rects = [r];
              }

            case 29:
              if (!(rects.length === 0)) {
                _context43.next = 33;
                break;
              }

              err = { reason: 'notfound', message: 'selector is not found. selector=' + selector };

              callback.apply(this, [err, null, selIdx, selectors]);
              return _context43.abrupt('continue', 54);

            case 33:
              rectIdx = 0;

            case 34:
              if (!(rectIdx < rects.length)) {
                _context43.next = 49;
                break;
              }

              rect = rects[rectIdx];

              if (!(rect.top >= meta.height || rect.left >= meta.width)) {
                _context43.next = 40;
                break;
              }

              _err = { reason: 'limitation', message: 'top of selector is over the limitation of height. selector=' + selector };

              callback.apply(this, [_err, null, selIdx, selectors]);
              return _context43.abrupt('continue', 46);

            case 40:
              if (meta.width < rect.left + rect.width) {
                rect.width = meta.width - rect.left;
              }
              if (meta.height < rect.top + rect.height) {
                rect.height = meta.height - rect.top;
              }

              _context43.next = 44;
              return _regenerator2.default.awrap(sharp(fullscreenBuffer).extract(rect).toBuffer());

            case 44:
              buffer = _context43.sent;

              callback.apply(this, [null, buffer, selIdx, selectors, rectIdx]);

            case 46:
              rectIdx++;
              _context43.next = 34;
              break;

            case 49:
              _context43.next = 54;
              break;

            case 51:
              _context43.prev = 51;
              _context43.t0 = _context43['catch'](17);

              callback.apply(this, [_context43.t0, null, selIdx, selectors]);

            case 54:
              selIdx++;
              _context43.next = 15;
              break;

            case 57:
              _context43.prev = 57;
              _context43.next = 60;
              return _regenerator2.default.awrap(emulation.reset());

            case 60:
              return _context43.finish(57);

            case 61:
            case 'end':
              return _context43.stop();
          }
        }
      }, null, this, [[13,, 57, 61], [17, 51]]);
    }
  }, {
    key: 'pdf',
    value: function pdf() {
      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var _ref2, data;

      return _regenerator2.default.async(function pdf$(_context44) {
        while (1) {
          switch (_context44.prev = _context44.next) {
            case 0:
              _context44.next = 2;
              return _regenerator2.default.awrap(this.client.Page.printToPDF(options));

            case 2:
              _ref2 = _context44.sent;
              data = _ref2.data;
              return _context44.abrupt('return', Buffer.from(data, 'base64'));

            case 5:
            case 'end':
              return _context44.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'startScreencast',
    value: function startScreencast(callback) {
      var _this11 = this;

      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      return _regenerator2.default.async(function startScreencast$(_context46) {
        while (1) {
          switch (_context46.prev = _context46.next) {
            case 0:
              _context46.next = 2;
              return _regenerator2.default.awrap(this.client.Page.screencastFrame(function _callee9(payload) {
                return _regenerator2.default.async(function _callee9$(_context45) {
                  while (1) {
                    switch (_context45.prev = _context45.next) {
                      case 0:
                        _context45.next = 2;
                        return _regenerator2.default.awrap(callback.apply(_this11, [payload]));

                      case 2:
                        _context45.next = 4;
                        return _regenerator2.default.awrap(_this11.client.Page.screencastFrameAck({ sessionId: payload.sessionId }));

                      case 4:
                      case 'end':
                        return _context45.stop();
                    }
                  }
                }, null, _this11);
              }));

            case 2:
              _context46.next = 4;
              return _regenerator2.default.awrap(this.client.Page.startScreencast(options));

            case 4:
            case 'end':
              return _context46.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'stopScreencast',
    value: function stopScreencast() {
      return _regenerator2.default.async(function stopScreencast$(_context47) {
        while (1) {
          switch (_context47.prev = _context47.next) {
            case 0:
              _context47.next = 2;
              return _regenerator2.default.awrap(this.client.Page.stopScreencast());

            case 2:
            case 'end':
              return _context47.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'requestWillBeSent',
    value: function requestWillBeSent(callback) {
      return _regenerator2.default.async(function requestWillBeSent$(_context48) {
        while (1) {
          switch (_context48.prev = _context48.next) {
            case 0:
              _context48.next = 2;
              return _regenerator2.default.awrap(this._checkStart());

            case 2:
              _context48.next = 4;
              return _regenerator2.default.awrap(this.client.Network.responseReceived(callback));

            case 4:
            case 'end':
              return _context48.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'inject',
    value: function inject(type, file) {
      var data, script, expr, style, _expr;

      return _regenerator2.default.async(function inject$(_context49) {
        while (1) {
          switch (_context49.prev = _context49.next) {
            case 0:
              _context49.next = 2;
              return _regenerator2.default.awrap(new Promise(function (resolve, reject) {
                fs.readFile(file, { encoding: 'utf-8' }, function (err, data) {
                  if (err) reject(err);
                  resolve(data);
                });
              }).catch(function (e) {
                throw e;
              }));

            case 2:
              data = _context49.sent;

              if (!(type === 'js')) {
                _context49.next = 9;
                break;
              }

              script = data.replace(/\\/g, '\\\\').replace(/'/g, '\\\'').replace(/(\r|\n)/g, '\\n');
              expr = '\n      {\n         let script = document.createElement(\'script\')\n         script.type = \'text/javascript\'\n         script.innerHTML = \'' + script + '\'\n         document.body.appendChild(script)\n      }\n      ';
              return _context49.abrupt('return', this.evaluate(expr));

            case 9:
              if (!(type === 'css')) {
                _context49.next = 15;
                break;
              }

              style = data.replace(/`/g, '\\`').replace(/\\/g, '\\\\'); // .replace(/(\r|\n)/g, ' ')

              _expr = '\n      {\n         let style = document.createElement(\'style\')\n         style.type = \'text/css\'\n         style.innerText = `\n        ' + style + '\n        `\n         document.head.appendChild(style)\n      }\n      ';
              return _context49.abrupt('return', this.evaluate(_expr));

            case 15:
              throw new Error('found invalid type.');

            case 16:
            case 'end':
              return _context49.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'emulate',
    value: function emulate(deviceName) {
      var device, platform;
      return _regenerator2.default.async(function emulate$(_context50) {
        while (1) {
          switch (_context50.prev = _context50.next) {
            case 0:
              _context50.next = 2;
              return _regenerator2.default.awrap(this._checkStart());

            case 2:
              if (this.emulateMode) {
                _context50.next = 6;
                break;
              }

              _context50.next = 5;
              return _regenerator2.default.awrap(this.evaluate('return navigator.userAgent'));

            case 5:
              this.userAgentBeforeEmulate = _context50.sent;

            case 6:
              device = devices[deviceName];
              _context50.next = 9;
              return _regenerator2.default.awrap(this.client.Emulation.setDeviceMetricsOverride({
                width: device.width,
                height: device.height,
                deviceScaleFactor: device.deviceScaleFactor,
                mobile: device.mobile,
                fitWindow: false,
                scale: device.pageScaleFactor
              }));

            case 9:
              platform = device.mobile ? 'mobile' : 'desktop';
              _context50.next = 12;
              return _regenerator2.default.awrap(this.client.Emulation.setTouchEmulationEnabled({ enabled: true, configuration: platform }));

            case 12:
              _context50.next = 14;
              return _regenerator2.default.awrap(this.userAgent(device.userAgent));

            case 14:
              this.currentEmulateDeviceName = deviceName;
              this.emulateMode = true;

            case 16:
            case 'end':
              return _context50.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'clearEmulate',
    value: function clearEmulate() {
      return _regenerator2.default.async(function clearEmulate$(_context51) {
        while (1) {
          switch (_context51.prev = _context51.next) {
            case 0:
              _context51.next = 2;
              return _regenerator2.default.awrap(this.client.Emulation.clearDeviceMetricsOverride());

            case 2:
              _context51.next = 4;
              return _regenerator2.default.awrap(this.client.Emulation.setTouchEmulationEnabled({ enabled: false }));

            case 4:
              if (!this.userAgentBeforeEmulate) {
                _context51.next = 7;
                break;
              }

              _context51.next = 7;
              return _regenerator2.default.awrap(this.userAgent(this.userAgentBeforeEmulate));

            case 7:
              this.emulateMode = false;
              this.currentEmulateDeviceName = null;

            case 9:
            case 'end':
              return _context51.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'blockUrls',
    value: function blockUrls(urls) {
      return _regenerator2.default.async(function blockUrls$(_context52) {
        while (1) {
          switch (_context52.prev = _context52.next) {
            case 0:
              _context52.next = 2;
              return _regenerator2.default.awrap(this._checkStart());

            case 2:
              _context52.next = 4;
              return _regenerator2.default.awrap(this.client.Network.setBlockedURLs({ urls: urls }));

            case 4:
            case 'end':
              return _context52.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'clearBrowserCache',
    value: function clearBrowserCache() {
      return _regenerator2.default.async(function clearBrowserCache$(_context53) {
        while (1) {
          switch (_context53.prev = _context53.next) {
            case 0:
              _context53.next = 2;
              return _regenerator2.default.awrap(this._checkStart());

            case 2:
              _context53.next = 4;
              return _regenerator2.default.awrap(this.client.Network.clearBrowserCache());

            case 4:
            case 'end':
              return _context53.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'setCookie',
    value: function setCookie(params) {
      return _regenerator2.default.async(function setCookie$(_context54) {
        while (1) {
          switch (_context54.prev = _context54.next) {
            case 0:
              _context54.next = 2;
              return _regenerator2.default.awrap(this._checkStart());

            case 2:
              _context54.next = 4;
              return _regenerator2.default.awrap(this.client.Network.setCookie(params));

            case 4:
            case 'end':
              return _context54.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'deleteCookie',
    value: function deleteCookie(name, url) {
      return _regenerator2.default.async(function deleteCookie$(_context55) {
        while (1) {
          switch (_context55.prev = _context55.next) {
            case 0:
              _context55.next = 2;
              return _regenerator2.default.awrap(this._checkStart());

            case 2:
              _context55.next = 4;
              return _regenerator2.default.awrap(this.client.Network.deleteCookie({ cookieName: name, url: url }));

            case 4:
            case 'end':
              return _context55.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'clearAllCookies',
    value: function clearAllCookies() {
      return _regenerator2.default.async(function clearAllCookies$(_context56) {
        while (1) {
          switch (_context56.prev = _context56.next) {
            case 0:
              _context56.next = 2;
              return _regenerator2.default.awrap(this._checkStart());

            case 2:
              _context56.next = 4;
              return _regenerator2.default.awrap(this.client.Network.clearBrowserCookies());

            case 4:
            case 'end':
              return _context56.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'getDOMCounters',
    value: function getDOMCounters() {
      return _regenerator2.default.async(function getDOMCounters$(_context57) {
        while (1) {
          switch (_context57.prev = _context57.next) {
            case 0:
              _context57.next = 2;
              return _regenerator2.default.awrap(this.client.Memory.getDOMCounters());

            case 2:
              return _context57.abrupt('return', _context57.sent);

            case 3:
            case 'end':
              return _context57.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'clearDataForOrigin',
    value: function clearDataForOrigin() {
      var origin = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
      var type = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'all';
      return _regenerator2.default.async(function clearDataForOrigin$(_context58) {
        while (1) {
          switch (_context58.prev = _context58.next) {
            case 0:
              if (!(origin === null)) {
                _context58.next = 4;
                break;
              }

              _context58.next = 3;
              return _regenerator2.default.awrap(this.evaluate(function (_) {
                return location.origin;
              }));

            case 3:
              origin = _context58.sent;

            case 4:
              _context58.next = 6;
              return _regenerator2.default.awrap(this.client.Storage.clearDataForOrigin({ origin: origin, storageTypes: type }));

            case 6:
              return _context58.abrupt('return', _context58.sent);

            case 7:
            case 'end':
              return _context58.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'scroll',
    value: function scroll(x, y) {
      return _regenerator2.default.async(function scroll$(_context59) {
        while (1) {
          switch (_context59.prev = _context59.next) {
            case 0:
              return _context59.abrupt('return', this._evaluateWithReplaces(function () {
                var dx = _1; // eslint-disable-line no-undef
                var dy = _2; // eslint-disable-line no-undef
                window.scrollTo(window.pageXOffset + dx, window.pageYOffset + dy);
              }, {}, { '_1': x, '_2': y }));

            case 1:
            case 'end':
              return _context59.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'scrollTo',
    value: function scrollTo(x, y) {
      return _regenerator2.default.async(function scrollTo$(_context60) {
        while (1) {
          switch (_context60.prev = _context60.next) {
            case 0:
              return _context60.abrupt('return', this._evaluateWithReplaces(function () {
                window.scrollTo(_1, _2); // eslint-disable-line no-undef
              }, {}, { '_1': x, '_2': y }));

            case 1:
            case 'end':
              return _context60.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'getBoundingClientRect',
    value: function getBoundingClientRect(selector) {
      var rect;
      return _regenerator2.default.async(function getBoundingClientRect$(_context61) {
        while (1) {
          switch (_context61.prev = _context61.next) {
            case 0:
              _context61.next = 2;
              return _regenerator2.default.awrap(this._evaluateWithReplaces(function () {
                var dom = document.querySelector('?');
                if (!dom) {
                  return null;
                }
                var r = dom.getBoundingClientRect();
                return { top: r.top, left: r.left, width: r.width, height: r.height };
              }, {}, { '?': escapeSingleQuote(selector) }));

            case 2:
              rect = _context61.sent;

              if (rect) {
                _context61.next = 5;
                break;
              }

              return _context61.abrupt('return', null);

            case 5:
              return _context61.abrupt('return', {
                top: Math.floor(rect.top),
                left: Math.floor(rect.left),
                width: Math.floor(rect.width),
                height: Math.floor(rect.height)
              });

            case 6:
            case 'end':
              return _context61.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: 'getBoundingClientRectAll',
    value: function getBoundingClientRectAll(selector) {
      var rects;
      return _regenerator2.default.async(function getBoundingClientRectAll$(_context62) {
        while (1) {
          switch (_context62.prev = _context62.next) {
            case 0:
              _context62.next = 2;
              return _regenerator2.default.awrap(this._evaluateWithReplaces(function () {
                var doms = document.querySelectorAll('?');
                return doms.map(function (dom) {
                  var r = dom.getBoundingClientRect();
                  return { top: r.top, left: r.left, width: r.width, height: r.height };
                });
              }, {}, { '?': escapeSingleQuote(selector) }));

            case 2:
              rects = _context62.sent;
              return _context62.abrupt('return', rects.map(function (rect) {
                return {
                  top: Math.floor(rect.top),
                  left: Math.floor(rect.left),
                  width: Math.floor(rect.width),
                  height: Math.floor(rect.height)
                };
              }));

            case 4:
            case 'end':
              return _context62.stop();
          }
        }
      }, null, this);
    }
  }, {
    key: '_checkStart',
    value: function _checkStart() {
      var startingUrl = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
      return _regenerator2.default.async(function _checkStart$(_context63) {
        while (1) {
          switch (_context63.prev = _context63.next) {
            case 0:
              if (!(this.client === null)) {
                _context63.next = 3;
                break;
              }

              _context63.next = 3;
              return _regenerator2.default.awrap(this.start(startingUrl));

            case 3:
            case 'end':
              return _context63.stop();
          }
        }
      }, null, this);
    }
  }], [{
    key: 'cleanup',
    value: function cleanup() {
      var copy, promises;
      return _regenerator2.default.async(function cleanup$(_context64) {
        while (1) {
          switch (_context64.prev = _context64.next) {
            case 0:
              copy = [].concat(instances);
              promises = copy.map(function (i) {
                return i.close();
              });
              _context64.next = 4;
              return _regenerator2.default.awrap(Promise.all(promises));

            case 4:
            case 'end':
              return _context64.stop();
          }
        }
      }, null, this);
    }
  }]);
  return Chromy;
}();

module.exports = Chromy;
